executor/common_linux.h

```diff
@@ -4535,3 +4610,280 @@ static volatile long syz_fuse_handle_req(volatile long a0, // /dev/fuse fd.
+		}
+
+		ret = set_interface_state(interface, 1);
+		if (ret < 0) {
+			debug("set_interface_state: set_interface_state failed for #%d, ret %d\n", device_id, ret);
+			goto error;
+		}
+
+		ret = nl80211_join_ibss(&tmp_msg, sock, nl80211_family_id, interface, &ibss_props);
+		if (ret < 0) {
+			debug("syz_80211_join_ibss: nl80211_join_ibss failed for #%d, ret %d\n", device_id, ret);
+			goto error;
+		}
+
+		if (await_up) {
+			do {
+				usleep(1000); // 1 ms
+				ret = get_ifla_operstate(&tmp_msg, interface);
+				if (ret < 0) {
+					debug("syz_80211_join_ibss: get_ifla_operstate failed for #%d, ret %d\n", device_id, ret);
+					goto error;
+				}
+			} while (ret != IF_OPER_UP);
+		}
+	}
+
+	close(sock);
+	return 0;
+error:
+	close(sock);
+	return -1;
+}
+
+#endif
+
+#if SYZ_EXECUTOR || __NR_syz_80211_inject_frame
+#include <linux/genetlink.h>
+#include <linux/if_ether.h>
+#include <linux/nl80211.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+
+#define HWSIM_ATTR_RX_RATE 5
+#define HWSIM_ATTR_SIGNAL 6
+#define HWSIM_ATTR_ADDR_RECEIVER 1
+#define HWSIM_ATTR_FRAME 3
+
+#define WIFI_MAX_INJECT_LEN 2048
+
+static int hwsim_register_socket(struct nlmsg* nlmsg, int sock, int hwsim_family)
+{
+	struct genlmsghdr genlhdr;
+	memset(&genlhdr, 0, sizeof(genlhdr));
+	genlhdr.cmd = HWSIM_CMD_REGISTER;
+	netlink_init(nlmsg, hwsim_family, 0, &genlhdr, sizeof(genlhdr));
+	int err = netlink_send(nlmsg, sock);
+	if (err < 0) {
+		debug("hwsim_register_device failed: %s\n", strerror(-err));
+		return -1;
+	}
+	return 0;
+}
+
+static int hwsim_inject_frame(struct nlmsg* nlmsg, int sock, int hwsim_family, uint8* mac_addr, uint8* data, int len)
+{
+	struct genlmsghdr genlhdr;
+	uint32 rx_rate = WIFI_DEFAULT_RX_RATE;
+	uint32 signal = WIFI_DEFAULT_SIGNAL;
+
+	memset(&genlhdr, 0, sizeof(genlhdr));
+	genlhdr.cmd = HWSIM_CMD_FRAME;
+	netlink_init(nlmsg, hwsim_family, 0, &genlhdr, sizeof(genlhdr));
+	netlink_attr(nlmsg, HWSIM_ATTR_RX_RATE, &rx_rate, sizeof(uint32));
+	netlink_attr(nlmsg, HWSIM_ATTR_SIGNAL, &signal, sizeof(uint32));
+	netlink_attr(nlmsg, HWSIM_ATTR_ADDR_RECEIVER, mac_addr, ETH_ALEN);
+	netlink_attr(nlmsg, HWSIM_ATTR_FRAME, data, len);
+	int err = netlink_send(nlmsg, sock);
+	if (err < 0) {
+		debug("inject_frame failed: %s\n", strerror(-err));
+		return -1;
+	}
+	return 0;
+}
+
+static long syz_80211_inject_frame(volatile long a0, volatile long a1, volatile long a2)
+{
+	uint8* mac_addr = (uint8*)a0;
+	uint8* buf = (uint8*)a1;
+	int buf_len = (int)a2;
+	struct nlmsg tmp_msg;
+
+	if (buf == NULL || buf_len < 0 || buf_len > WIFI_MAX_INJECT_LEN)
+		return -1; // restricting input in order to avoid unnecessary segfaults on memcpy
+
+	int sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_GENERIC);
+	if (sock < 0) {
+		debug("syz_80211_join_ibss: socket creation failed, errno %d\n", errno);
+		return -1;
+	}
+
+	int hwsim_family_id = netlink_query_family_id(&tmp_msg, sock, "MAC80211_HWSIM");
+	int ret = hwsim_register_socket(&tmp_msg, sock, hwsim_family_id);
```

Comment:
This registers netlink port id somewhere in hswim subsystem. But we always use 0 port id. Can't we do this once in initialize_wifi_devices?

Reply:
No, we cannot do that because this command registers a specific socket. Unregistration happens automatically once the socket is closed.

Also, we cannot create one socket, register it, and use all the time, because this prevents mac80211_hwsim from emulating perfect medium. It disrupts normal networking activities.